ethereum-rpc/src/constructors/express/routes/api/v1/tokens.js

```diff
@@ -1,43 +1,178 @@
-const axios = require('axios')
-const router = require('express').Router()
+const axios = require("axios");
+const router = require("express").Router();
 
 
-module.exports = ({ models, ethereum: { web3, buildContract, compiler }, ...context }) => {
+const prefixContractAddress = address => address.startsWith('0x') ? address : `0x${address}`
 
-  router.post('/', async (req, res, next) => {
+module.exports = ({
+  models,
+  ethereum: { web3, buildContract, compiler, abi },
+  ...context
+}) => {
+  router.post("/", async (req, res, next) => {
     try {
       const {
-        network = 'mainnet',
-        type = 'UNISWAP',
+        network = "mainnet",
+        type = "UNISWAP",
         amount = 0,
-        sender = '',
-        recipient = '',
-        tokenContractAddress = '',
-        privateKey = '',
+        sender = "",
+        recipient = "",
+        tokenContractAddress = "",
+        privateKey = "",
         sendOptions = {
           gasLimit: web3.utils.toHex(100000),
           gasPrice: web3.utils.toHex(50e9),
         },
         token = {},
         ...body
-      } = req.body
+      } = req.body;
 
       const result = await compiler.deployContract({
         token,
         type,
         sender,
         privateKey,
         sendOptions,
-      })
+      });
 
-      console.log('result', result)
+      console.log("result", result);
 
-      return res.status(200).json(result.deployTransaction)
+      return res.status(200).json(result.deployTransaction);
     } catch (e) {
-      console.error(e)
-      return res.status(500).json({ errors: [e] })
+      console.error(e);
+      return res.status(500).json({ errors: [e] });
     }
-  })
+  });
 
-  return router
-}
+  // balance functionality for erc20 and erc721
+
+  router.get("/balance/:address", async (req, res, next) => {
+    try {
+      const { address = "*" } = req.params;
+      let { tokenAddress = "", type = "erc20" } = { ...req.query };
+      const tokenContractAddress = prefixContractAddress(
+        tokenAddress.toString("hex")
+      );
+      if (tokenAddress.length < 42 || tokenAddress.length > 42) {
+        return res.status(500).json({ error: "invalid contract address" });
+      }
+      let tokenType = type.toString();
+
+      if (!(tokenType == "erc20" || tokenType == "erc721")) {
+        return res.status(500).json({ error: "invalid token type" });
+      }
+      const contractAbi = tokenType == "erc20" ? abi.APIS_ : abi.NFT_7;
```

Comment:
Does `abi.APIS_` resolve? Also, why is the file name `NFT_7`, as opposed to `ERC_721` for example?

Reply:
Nevermind! This is my fault -- let's fix the logic in ethereum/index.js which does a substring on the filename. Let's use the actual filename without .sol extension as the key. And let's make sure our filenames are self-documenting.